<launch>
  <arg name="use_fcn" default="true" />
  <arg name="stereo" default="true" />

  <!-- hand mounted camera -->
  <group if="$(arg stereo)">
    <include file="$(find jsk_arc2017_baxter)/launch/setup/include/stereo_astra_hand.launch" />
  </group>
  <group unless="$(arg stereo)">
    <include file="$(find jsk_arc2017_baxter)/launch/setup/include/astra_hand.launch" />
  </group>

  <!-- weight scale -->
  <include file="$(find jsk_arc2017_baxter)/launch/setup/include/scale.launch" />

  <arg name="LEFT_CAMERA_INFO" if="$(arg stereo)" value="/left_hand_camera/fused/rgb/camera_info" />
  <arg name="LEFT_INPUT_CLOUD" if="$(arg stereo)" value="/left_hand_camera/fused/depth_registered/points" />
  <arg name="LEFT_INPUT_IMAGE" if="$(arg stereo)" value="/left_hand_camera/left/rgb/image_rect_color" />
  <arg name="LEFT_MANAGER" if="$(arg stereo)" value="/left_hand_camera/left/left_nodelet_manager" />
  <arg name="RIGHT_CAMERA_INFO" if="$(arg stereo)" value="/right_hand_camera/fused/rgb/camera_info" />
  <arg name="RIGHT_INPUT_CLOUD" if="$(arg stereo)" value="/right_hand_camera/fused/depth_registered/points" />
  <arg name="RIGHT_INPUT_IMAGE" if="$(arg stereo)" value="/right_hand_camera/left/rgb/image_rect_color" />
  <arg name="RIGHT_MANAGER" if="$(arg stereo)" value="/right_hand_camera/left/left_nodelet_manager" />

  <arg name="LEFT_CAMERA_INFO" unless="$(arg stereo)" value="/left_hand_camera/depth_registered/camera_info" />
  <arg name="LEFT_INPUT_CLOUD" unless="$(arg stereo)" value="/left_hand_camera/depth_registered/points" />
  <arg name="LEFT_INPUT_IMAGE" unless="$(arg stereo)" value="/left_hand_camera/rgb/image_rect_color" />
  <arg name="LEFT_MANAGER" unless="$(arg stereo)" value="/left_hand_camera/left_hand_camera_nodelet_manager" />
  <arg name="RIGHT_CAMERA_INFO" unless="$(arg stereo)" value="/right_hand_camera/depth_registered/camera_info" />
  <arg name="RIGHT_INPUT_CLOUD" unless="$(arg stereo)" value="/right_hand_camera/depth_registered/points" />
  <arg name="RIGHT_INPUT_IMAGE" unless="$(arg stereo)" value="/right_hand_camera/rgb/image_rect_color" />
  <arg name="RIGHT_MANAGER" unless="$(arg stereo)" value="/right_hand_camera/right_hand_camera_nodelet_manager" />

  <group ns="left_hand_camera">
    <node name="bbox_array_to_bbox"
          pkg="nodelet" type="nodelet"
          args="load jsk_pcl_utils/BoundingBoxArrayToBoundingBox $(arg LEFT_MANAGER)">
      <remap from="~input" to="/transformable_tote_markers/output/boxes" />
      <rosparam>
        index: 0
      </rosparam>
    </node>
    <node name="attention_clipper_tote"
          pkg="nodelet" type="nodelet"
          args="load jsk_pcl/AttentionClipper $(arg LEFT_MANAGER)">
      <remap from="~input/points" to="$(arg LEFT_INPUT_CLOUD)" />
      <remap from="~input/box" to="bbox_array_to_bbox/output" />
      <rosparam>
        use_multiple_attention: false
      </rosparam>
    </node>
    <node name="extract_indices_tote"
          pkg="nodelet" type="nodelet"
          args="load jsk_pcl/ExtractIndices $(arg LEFT_MANAGER)">
      <remap from="~input" to="$(arg LEFT_INPUT_CLOUD)" />
      <remap from="~indices" to="attention_clipper_tote/output/point_indices" />
      <rosparam>
        keep_organized: true
        approximate_sync: true
        max_queue_size: 100
      </rosparam>
    </node>
    <group if="$(arg use_fcn)">
      <include file="$(find jsk_arc2017_common)/launch/object_segmentation_3d.launch">
        <arg name="NODELET_MANAGER" value="$(arg LEFT_MANAGER)" />
        <arg name="INPUT_IMAGE" value="$(arg LEFT_INPUT_IMAGE)"/>
        <arg name="INPUT_CLOUD" value="extract_indices_tote/output"/>
        <arg name="INPUT_CANDIDATES" value="candidates_publisher/output/candidates" />
        <arg name="SELECT_LARGEST" value="true" />
      </include>
    </group>
    <group unless="$(arg use_fcn)">
      <node name="supervoxel_segmentation"
            pkg="nodelet" type="nodelet"
            args="load jsk_pcl/SupervoxelSegmentation $(arg LEFT_MANAGER)">
        <remap from="~input" to="/left_hand_camera/extract_indices_tote/output" />
        <rosparam>
          color_importance: 0.2
          spatial_importance: 0.4
          normal_importance: 1.0
          voxel_resolution: 0.01
          seed_resolution: 0.1
        </rosparam>
      </node>
      <node name="cluster_indices_decomposer_target"
            pkg="nodelet" type="nodelet"
            args="load jsk_pcl/ClusterPointIndicesDecomposer $(arg LEFT_MANAGER)">
        <remap from="~input" to="/left_hand_camera/supervoxel_segmentation/output/cloud" />
        <remap from="~target" to="/left_hand_camera/supervoxel_segmentation/output/indices" />
        <rosparam>
          queue_size: 100
        </rosparam>
      </node>
    </group>

    <!-- weight object classifier -->
    <node name="weight_candidates_refiner"
          pkg="jsk_arc2017_common" type="weight_candidates_refiner.py">
      <remap from="~input/candidates" to="candidates_publisher/output/candidates" />
      <rosparam ns="object_weights" file="$(find jsk_arc2017_common)/data/objects/weight.yaml" />
      <rosparam>
        approximate_sync: true
        queue_size: 10
        input_topics:
          - /scale0/output
          - /scale1/output
          - /scale2/output
          - /scale3/output
        error: 10.0
      </rosparam>
    </node>
  </group>  <!-- ns="left_hand_camera" -->

  <group ns="right_hand_camera">
    <node name="bbox_array_to_bbox"
          pkg="nodelet" type="nodelet"
          args="load jsk_pcl_utils/BoundingBoxArrayToBoundingBox $(arg RIGHT_MANAGER)">
      <remap from="~input" to="/transformable_tote_markers/output/boxes" />
      <rosparam>
        index: 1
      </rosparam>
    </node>
    <node name="attention_clipper_tote"
          pkg="nodelet" type="nodelet"
          args="load jsk_pcl/AttentionClipper $(arg RIGHT_MANAGER)">
      <remap from="~input/points" to="$(arg RIGHT_INPUT_CLOUD)" />
      <remap from="~input/box" to="bbox_array_to_bbox/output" />
      <rosparam>
        use_multiple_attention: false
      </rosparam>
    </node>
    <node name="extract_indices_tote"
          pkg="nodelet" type="nodelet"
          args="load jsk_pcl/ExtractIndices $(arg RIGHT_MANAGER)">
      <remap from="~input" to="$(arg RIGHT_INPUT_CLOUD)" />
      <remap from="~indices" to="attention_clipper_tote/output/point_indices" />
      <rosparam>
        keep_organized: true
        approximate_sync: true
        max_queue_size: 100
      </rosparam>
    </node>
    <group if="$(arg use_fcn)">
      <include file="$(find jsk_arc2017_common)/launch/object_segmentation_3d.launch">
        <arg name="NODELET_MANAGER" value="$(arg RIGHT_MANAGER)" />
        <arg name="INPUT_IMAGE" value="$(arg RIGHT_INPUT_IMAGE)"/>
        <arg name="INPUT_CLOUD" value="extract_indices_tote/output"/>
        <arg name="INPUT_CANDIDATES" value="candidates_publisher/output/candidates" />
        <arg name="SELECT_LARGEST" value="true" />
      </include>
    </group>
    <group unless="$(arg use_fcn)">
      <node name="supervoxel_segmentation"
            pkg="nodelet" type="nodelet"
            args="load jsk_pcl/SupervoxelSegmentation $(arg RIGHT_MANAGER)">
        <remap from="~input" to="/right_hand_camera/extract_indices_tote/output" />
        <rosparam>
          color_importance: 0.2
          spatial_importance: 0.4
          normal_importance: 1.0
          voxel_resolution: 0.01
          seed_resolution: 0.1
        </rosparam>
      </node>
      <node name="cluster_indices_decomposer_target"
            pkg="nodelet" type="nodelet"
            args="load jsk_pcl/ClusterPointIndicesDecomposer $(arg RIGHT_MANAGER)">
        <remap from="~input" to="/right_hand_camera/supervoxel_segmentation/output/cloud" />
        <remap from="~target" to="/right_hand_camera/supervoxel_segmentation/output/indices" />
        <rosparam>
          queue_size: 100
        </rosparam>
      </node>
    </group>

    <!-- weight object classifier -->
    <node name="weight_candidates_refiner"
          pkg="jsk_arc2017_common" type="weight_candidates_refiner.py">
      <remap from="~input/candidates" to="candidates_publisher/output/candidates" />
      <rosparam ns="object_weights" file="$(find jsk_arc2017_common)/data/objects/weight.yaml" />
      <rosparam>
        approximate_sync: true
        queue_size: 10
        input_topics:
          - /scale0/output
          - /scale1/output
          - /scale2/output
          - /scale3/output
        error: 10.0
      </rosparam>
    </node>
  </group>  <!-- ns="right_hand_camera" -->

 </launch>
